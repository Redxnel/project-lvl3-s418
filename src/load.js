import { promises as fs } from 'fs';
import axios from 'axios';
import cheerio from 'cheerio';
import path from 'path';
import { resolve } from 'url';
import debug from 'debug';
import buildName from './buildName';

const log = debug('page-loader');

const getDataFromUrl = uri => axios({
  method: 'get',
  url: uri,
  responseType: 'arrayBuffer',
})
  .then(response => response.data);

const tags = {
  link: 'href',
  img: 'src',
  script: 'src',
};

const isLocalLink = (link) => {
  const substr = link.substr(0, 4);
  return substr !== 'http';
};

const generateNameLink = (link, filepath) => `${filepath}/${link.split('/').filter(el => el).join('-')}`;

const getLinksAndModifyHtml = (html, filepath) => {
  const $ = cheerio.load(html);
  const links = Object.keys(tags).reduce((acc, tag) => {
    const findedLink = $(tag).map((i, element) => {
      const link = $(element).attr(tags[tag]) || '';
      const newLink = isLocalLink(link) ? generateNameLink(link, filepath) : link;
      $(element).attr(tags[tag], newLink);
      return isLocalLink(link) ? link : '';
    }).get();
    return [...acc, ...findedLink].filter(elem => elem);
  }, []);
  log('New html created success!')
  return { links, html: $.html({ decodeEntities: false }) };
};

export default (page, filepath) => {
  let dirpath = '';
  let newHtml = '';
  const directoryForResource = buildName(page, filepath, '_files');
  return fs.mkdir(directoryForResource)
    .then(() => getDataFromUrl(page))
    .then(data => getLinksAndModifyHtml(data, directoryForResource))
    .then((content) => {
      newHtml = content.html;
      const linksFromPage = content.links;
      return Promise.all(linksFromPage.map(link => getDataFromUrl(resolve(page, link))
        .then((data) => {
          const updatedLink = generateNameLink(link, directoryForResource);
          const pathToWrite = path.join(updatedLink);
          log(`${link} updated to ${updatedLink}`);
          return fs.writeFile(pathToWrite, data, 'utf-8');
        })));
    })
    .then(() => {
      dirpath = buildName(page, filepath, '.html');
      return fs.writeFile(dirpath, newHtml, 'utf-8');
    })
    .then(() => {
      log(`page written on disc in ${dirpath}`);
      return dirpath;
    });
};
