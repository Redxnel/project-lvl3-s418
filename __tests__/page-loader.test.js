import nock from 'nock';
import httpAdapter from 'axios/lib/adapters/http';
import { promises as fs } from 'fs';
import os from 'os';
import path from 'path';
import axios from 'axios';
import load from '../src';

const host = 'http://localhost';

axios.defaults.adapter = httpAdapter;

let tmpdir = '';
let dataCss = '';
let dataHtml = '';

beforeEach(async () => {
  tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'test'));
  dataCss = await fs.readFile('./__tests__/__fixtures__/style.css', 'utf-8');
  dataHtml = await fs.readFile('./__tests__/__fixtures__/resume.html', 'utf-8');
});

describe('Test', () => {
  it('Download simple page', async () => {
    nock(host)
      .get('/test')
      .reply(200, 'test data');

    await load(`${host}/test`, tmpdir);
    const recievedData = await fs.readFile(path.join(tmpdir, 'localhost-test.html'), 'utf-8');
    return expect(recievedData).toBe('<html><head></head><body>test data</body></html>');
  });

  it('Download page and resource', async () => {
    nock(host)
      .get('/')
      .replyWithFile(200, `${__dirname}/__fixtures__/resume.html`);

    nock(host)
      .get('/style.css')
      .replyWithFile(200, `${__dirname}/__fixtures__/style.css`);

    nock(host)
      .get('/resume.png')
      .replyWithFile(200, `${__dirname}/__fixtures__/resume.png`);

    await load(host, tmpdir);
    const recievedData = await fs.readFile(path.join(tmpdir, 'localhost.html'), 'utf-8');
    const recievedCss = await fs.readFile(path.join(tmpdir, 'localhost_files/style.css'), 'utf-8');
    const quantityFileInTmpDir = await fs.readdir(path.join(tmpdir, 'localhost_files'));
    expect(recievedData).not.toMatch(dataHtml);
    expect(recievedCss).toMatch(dataCss);
    expect(quantityFileInTmpDir.length).toBe(2);
  });
});
