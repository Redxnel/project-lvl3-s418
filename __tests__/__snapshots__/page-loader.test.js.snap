// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Errors 404 1`] = `"Request failed with status code 404"`;

exports[`Test Errors directory 1`] = `"ENOENT: no such file or directory, mkdir 'var/localhost-test_files'"`;